File: ruby/proverb/_iterations/1/proverb.rb
class Proverb
  def initialize(*input, qualifier: "")
    @input = input
    @qualifier = qualifier
  end

  def to_s
    (@input.zip(@input.drop(1)).map do |want, lost|
      "For want of a #{want} the #{lost} was lost."
    end.take(@input.size - 1) + [last_line]).join("\n")
  end

  private

  def last_line
    "And all for the want of a #{@qualifier}#{@qualifier.empty? ? '' : ' '}#{@input.first}."
  end
end

File: clojure/proverb/_iterations/1/src/proverb.clj
(ns proverb
  (:require [clojure.string :as string]))

(defn- line [[want lost]]
  (str "For want of a " want " the " lost " was lost."))

(defn recite [[first-piece :as pieces]]
  (if-not first-piece
    ""
    (->> pieces
         (partition 2 1)
         (mapv line)
         (#(conj % (str "And all for the want of a " first-piece ".")))
         (string/join "\n"))))
